### ISSUE TITLE
defining how DID can be used as user's indetifier

### ISSUE THREAD

**Issue by Sakurann** [oauth-wg/oauth-sd-jwt-vc#205]
There are use-cases that would like to use IETF SD-JWT VC draft over W3C VCDM draft, while using DID as user's/subject's identifier. Currently, SD-JWT VC mandates `cnf` claim to be present. It would be very helpful to define how a DID can be used with a `cnf` claim. (as keeping `cnf` claim as mandatory is probably better for the interoperability)

**Comment by awoie** [comment]
Yes, we had that discussion recently. It will be something as follows:

```
"cnf":{
  "did":{ "id": "did:example:1234" }
}
```

**Comment by Sakurann** [comment]
Where should cnf.did be defined?

**Comment by bc-pi** [comment]
My intuition is that sd-jwt-vc is not an appropriate place to define and register a new DID based JWT Confirmation Method. I'm not sure where it should be done but not here. 

Meanwhile [cnf.kid](https://www.rfc-editor.org/rfc/rfc7800.html#section-3.4) seems sufficient for the case.

**Comment by Sakurann** [comment]
cnf.kid instead of cnf.did? why would we need to defined cnf.did if we think cnf.kid is sufficient?

**Comment by bc-pi** [comment]
I guess that was kinda what I was implying/suggesting. That cnf.kid is sufficient so cnf.did doesn't need to be defined.  

The interpretation of the kid value is application specific (per RFC7800) and can convey the same info as Oliver's example:

```
      "cnf":{
         "kid": "did:example:1234"
       }
```

**Comment by awoie** [comment]
I fully agree that it is not in scope of SD-JWT VC to define the did cnf method. However, I think this would be the cleaner approach. I would expect clients using this cnf method to register a did cnf method. kid does not really help with interoperability.

**Comment by Sakurann** [comment]
> kid does not really help with interoperability.
Why? either relative of absolute DID URL, it is being used as a `kid` value already.

**Comment by awoie** [comment]
> > kid does not really help with interoperability.
> 
> Why? either relative of absolute DID URL, it is being used as a `kid` value already.

There is nothing wrong with this but I would assume that most implementations won't test if `kid` is a DID URL. To make it explicit people could just use a dedicated DID `cnf` method.

**Comment by awoie** [comment]
@Sakurann is there anything where you believe we should update the SD-JWT VC specification?

**Comment by Sakurann** [comment]
> is there anything where you believe we should update the SD-JWT VC specification?

one sentence saying that "when key is passed by reference (assuming Briand would not want to explicitly mention DIDs in the spec), cnf.kid should be used" would be helpful, so that people reading the specification know that SD-JWT VC can be used with DIDs too.

**Comment by bc-pi** [comment]
I took a try at the following proposed new text to replace [cnf here](https://www.ietf.org/archive/id/draft-ietf-oauth-sd-jwt-vc-01.html#section-3.2.2.2-3.5.1) that attempts to be inclusive of opinions expressed in this issue and also addresses issue #196 :

>  cnf
> - OPTIONAL. Contains the confirmation method identifying the proof of possession key as defined in [[RFC7800](https://drafts.oauth.net/oauth-sd-jwt-vc/draft-ietf-oauth-sd-jwt-vc.html#RFC7800)]. This claim MUST be present when cryptographic Key Binding is to be supported. It is RECOMMENDED that this contains a JWK as defined in Section 3.2 of [[RFC7800](https://drafts.oauth.net/oauth-sd-jwt-vc/draft-ietf-oauth-sd-jwt-vc.html#RFC7800)]. When the key is to be conveyed by reference (e.g., via a key thumbprint or DID [[W3C.DID](https://drafts.oauth.net/oauth-sd-jwt-vc/draft-ietf-oauth-sd-jwt-vc.html#W3C.DID)]), other JWT confirmation methods MAY be used, such as a key identifier defined in Section 3.4 of [[RFC7800](https://drafts.oauth.net/oauth-sd-jwt-vc/draft-ietf-oauth-sd-jwt-vc.html#RFC7800)].   For proof of cryptographic Key Binding, the Key Binding JWT in the presentation of the SD-JWT MUST be signed by the key identified in this claim.

**Comment by Sakurann** [comment]
+1 to the proposed text - thank you, Brian!

**Comment by awoie** [comment]
I'd be ok with this change although we need to adjust the language a bit. 

Option 1:
- We should restrict this (`kid` as `cnf`) to being a DID URL instead of allowing `kid` being a DID. This is because a DID URL refers to one specific key whereas a DID is associated with multiple keys. I don't want to encourage implementations where the verifier has to try to verify the signature with potentially multiple keys unsuccessfully until they found one that matches. The proposed language would result in that if we don't limit it to DID URLs.

Option 2:
- We should adapt the language as follows: `For proof of cryptographic Key Binding, the Key Binding JWT in the presentation of the SD-JWT MUST be signed by a key identified or referenced by this claim`. A DID is not a key and needs resolution. 

I would prefer Option 1 and recommend that if having a DID as a being included in the `cnf` value is required, defining a dedicated `cnf` method for that would be needed which is a much cleaner approach. Note that I think that using DID URL for the `kid` value in the signed KB-JWT/Issuer-signed JWT is totally fine. I don't think we need to put this recommendation into the spec. 

There is a reason why there is a dedicated `x5c` `cnf` method instead of putting the base64url-DER-encoded certificate in the `cnf` claim.

**Comment by paulbastian** [comment]
I agree with Oliver. 

Kid is described as application specific and having a dedicated cnf.did method gives it a more universal touch that may be beneficial for interoperability.

**Comment by bc-pi** [comment]
I think the most beneficial thing for interoperability would be always and only using `cnf.jwk` for key binding. But that's a digression... 

This document is not going to define a dedicated cnf.did method.

The sentence, "When the key is to be conveyed by reference (e.g., via a key thumbprint or DID [W3C.DID]), other JWT confirmation methods MAY be used, such as a key identifier defined in Section 3.4 of [RFC7800]." in the proposed text [above](https://github.com/oauth-wg/oauth-sd-jwt-vc/issues/205#issuecomment-1947438330) is a just `MAY` to say that a `cnf.kid` or to-be-defined `cnf.did` could be used to conveyed the key binding key by reference.   I'd honestly rather not have it that sentence at all (note that using `cnf.kid` or to-be-defined `cnf.did` isn't prohibited now it's just not mentioned) but it was my attempt to accommodate the ask by @Sakurann.  Saying DID URL rather than just DID in that sentence would be okay for me too. But I don't think much more in this area would be appropriate here.

**Comment by cre8** [comment]
It depends how we define "proof of possession key" in context of a DID document, where multiple keys can be used to proof the identity. In case of natural persons when they prefer to use dids, it's more likely they use `did:key` or `did:jwk`. But since SD-JWT-VCs can also be issued to a legal person that can be using `did:web`, `did:ethr` it is able to register multiple keys.

Key rotation in this scenario are a good security feature because they allow me to update my crypto material without issuing me new credentials that are directly bound to it. On the other hand I am unable to insert security mechanism like hashlinks since I want to point not directly to one specific public key, but to one having specific rights.

To implement this, we can use the [Representation of a URL for a Proof-of-Possession Key](https://www.rfc-editor.org/rfc/rfc7800.html#section-3.5):
```

     {
      "iss": "https://server.example.com",
      "sub": "17760704",
      "aud": "https://client.example.org",
      "exp": 1440804813,
      "cnf":{
        "jku": "did:exmaple.com",
        "kid": "key123"
       }
     }
```
JKU has to be an URI, DID is one. We do not need to add the fragment since all elements inside a did document are addressed with it.

**Comment by awoie** [comment]
> It depends how we define "proof of possession key" in context of a DID document, where multiple keys can be used to proof the identity. In case of natural persons when they prefer to use dids, it's more likely they use `did:key` or `did:jwk`. But since SD-JWT-VCs can also be issued to a legal person that can be using `did:web`, `did:ethr` it is able to register multiple keys.
> 
> Key rotation in this scenario are a good security feature because they allow me to update my crypto material without issuing me new credentials that are directly bound to it. On the other hand I am unable to insert security mechanism like hashlinks since I want to point not directly to one specific public key, but to one having specific rights.

Yes, this makes sense. However, cnf.did might be more appropriate to encode this information.

**Comment by cre8** [comment]
> > It depends how we define "proof of possession key" in context of a DID document, where multiple keys can be used to proof the identity. In case of natural persons when they prefer to use dids, it's more likely they use `did:key` or `did:jwk`. But since SD-JWT-VCs can also be issued to a legal person that can be using `did:web`, `did:ethr` it is able to register multiple keys.
> > Key rotation in this scenario are a good security feature because they allow me to update my crypto material without issuing me new credentials that are directly bound to it. On the other hand I am unable to insert security mechanism like hashlinks since I want to point not directly to one specific public key, but to one having specific rights.
> 
> Yes, this makes sense. However, cnf.did might be more appropriate to encode this information.

Yeah, since we can not reuse this approach from the spec since it has a defined response.
```
"cnf":{
  "jku": "did:exmaple.com",
  "kid": "key123"
}
```
Wouldn't it be the best way to extend the RFC for CNF to include the did resolution as an option? Of course this is more work, but then we can reference to a standard

**Comment by paulbastian** [comment]
> > > It depends how we define "proof of possession key" in context of a DID document, where multiple keys can be used to proof the identity. In case of natural persons when they prefer to use dids, it's more likely they use `did:key` or `did:jwk`. But since SD-JWT-VCs can also be issued to a legal person that can be using `did:web`, `did:ethr` it is able to register multiple keys.
> > > Key rotation in this scenario are a good security feature because they allow me to update my crypto material without issuing me new credentials that are directly bound to it. On the other hand I am unable to insert security mechanism like hashlinks since I want to point not directly to one specific public key, but to one having specific rights.
> > 
> > 
> > Yes, this makes sense. However, cnf.did might be more appropriate to encode this information.
> 
> Yeah, since we can not reuse this approach from the spec since it has a defined response.
> 
> ```
> "cnf":{
>   "jku": "did:exmaple.com",
>   "kid": "key123"
> }
> ```
> 
> Wouldn't it be the best way to extend the RFC for CNF to include the did resolution as an option? Of course this is more work, but then we can reference to a standard

`jku` can not be used in my opinion.

RFC7800 says:

> This is done using the "jku" member.  Its  value is a URI [[RFC3986](https://www.rfc-editor.org/rfc/rfc3986)] that refers to a resource for a set of JSON-encoded public keys represented as a [JWK](https://www.rfc-editor.org/rfc/rfc7800.html#ref-JWK) Set [JWK], one of which is the proof-of-possession key.

I do not think that this matches to (most) DID documents.

**Comment by awoie** [comment]
> > kid does not really help with interoperability.
> 
> Why? either relative of absolute DID URL, it is being used as a `kid` value already.

We should have further restrictions on the DID URL. a `kid` containing a DID URL that is a DID would resolve to multiple keys potentially, which would cause ambiguity. 

I made a mistake previously when I said using a DID URL would solve the problem since a DID is also a DID URL:
> did-url = did path-abempty [ "?" query ] [ "#" fragment ]

whereas path-abempty can be literally completely empty which allows a DID to be a DID URL which makes sense because it is the location of the DID Document.

We would need to say that kid must be a DID URL that when dereferenced results in a single key (ideally JWK).

**Comment by Sakurann** [comment]
I think the text Brian suggested originally is sufficient: https://github.com/oauth-wg/oauth-sd-jwt-vc/issues/205#issuecomment-1947438330

(at least initially) please keep the option to use cnf.kid, as limiting the use of DIDs only with cnf.did could be a significant blocker for some implementations to implement sd-jwt vc. especially if sd-jwt vc will not define cnf.did (and I won't even attempt to try convince Brian to do so).

I also think it is probably better to say DID URL as opposed to just DID.

**Comment by awoie** [comment]
> I think the text Brian suggested originally is sufficient: [#205 (comment)](https://github.com/oauth-wg/oauth-sd-jwt-vc/issues/205#issuecomment-1947438330)
> 
> (at least initially) please keep the option to use cnf.kid, as limiting the use of DIDs only with cnf.did could be a significant blocker for some implementations to implement sd-jwt vc. especially if sd-jwt vc will not define cnf.did (and I won't even attempt to try convince Brian to do so).
> 
> I also think it is probably better to say DID URL as opposed to just DID.

Why is it a blocker if they use cnf.did instead of cnf.kid?

**Comment by awoie** [comment]
After some discussions, we believe the best approach is to define a dedicated `cnf` method for DIDs through a very small spec.

**Comment by TimoGlastra** [comment]
As having implemented DID holder binding for SD-JWT VCs while this discussion has been ongoing, we've decided to add the didUrl (pointing to the specific key that has been verified before issuance) in the `kid` value of the `cnf` claim.

So it will look something like:

```
{
   "cnf": {
       "kid": "did:example:123#key-id"
   }
}
```

Not sure if relevant for this discussion, but to show some implementers feedback on something that we have working today.

Happy to change our implementation on what comes out of this discussion.

**Comment by awoie** [comment]
> As having implemented DID holder binding for SD-JWT VCs while this discussion has been ongoing, we've decided to add the didUrl (pointing to the specific key that has been verified before issuance) in the `kid` value of the `cnf` claim.
> 
> So it will look something like:
> 
> ```
> {
>    "cnf": {
>        "kid": "did:example:123#key-id"
>    }
> }
> ```
> 
> Not sure if relevant for this discussion, but to show some implementers feedback on something that we have working today.
> 
> Happy to change our implementation on what comes out of this discussion.

While I cannot prevent anybody from using this, I would find it much cleaner to use a dedicated DID claim for this for interoperability reasons. It should be fairly easy to write a DID `cnf` IETF draft:

```
{
   "cnf": {
       "did": { "did_url": "did:example:123#key-id" }
   }
}
```

**Comment by TimoGlastra** [comment]
As i mentioned in my post, we're happy to change our implementation on what comes out of this discussion.

But until then re-using existing fields is a lot easier to integrate. `kid` been used for did key identifiers for multiple years already

**Comment by danielfett** [comment]
We discussed this on our editors' call and agreed that this method should be defined in a separate spec.

**Comment by peacekeeper** [comment]
I don't really have a strong opinion on whether "kid" or "jku" a new DID-based "cnf" method should be used.

In any case however, it would still make sense to have language such as what has been proposed by @bc-pi in https://github.com/oauth-wg/oauth-sd-jwt-vc/issues/205#issuecomment-1947438330, considering existing implementations and use cases. Also, an example using a DID as a user's identifier would be useful.

**Comment by bc-pi** [comment]
As previously discussed, a DID based confirmation method should be defined in a separate spec. I'd encourage interested parties to pursue that.

**Comment by cre8** [comment]
@peacekeeper @bc-pi would it help to add to the spec that sd-jwt-vc will support all listed [JWT Confirmation Methods](https://www.iana.org/assignments/jwt/jwt.xhtml).

So the way to go would be to write the spec for `DID` for confirmation method, add it to the list (procedure listed there) and then it can be used here. 

Since key binding has no reference to a specific confirmation method, it should have no impact.

**Comment by Sakurann** [comment]
I originally opened this issue, and I agree with closing it. I think a lot has evolved since I opened this issue and I am not supportive of sd-jwt vc spec mentioning DIDs. If people still want DIDs, cnf.did should be defined (i have been saying it for few years), which would allow DIDs to be used because sd-jwt vc defines the usage of a cnf claim.
